docs/spec/fastsync/fastsync.tla

```diff
@@ -0,0 +1,661 @@
+    blocksRequested                          \* set of BlockRequests received that are not answered yet
+  ]
+  *)
+
+ \* the variables for the network and scheduler
+ VARIABLES
+  turn,                                     \* who is taking the turn: "Peers" or "Node"
+  inMsg,                                    \* a node receives message by this variable
+  outMsg                                    \* a node sends a message by this variable
+
+
+(* the variables of the node *)
+nvars == <<state, blockPool>>
+
+\* Control messages
+ControlMsgs ==
+    [type: {"addPeer"}, peerId: AllPeerIds]
+        \union
+    [type: {"removePeer"}, peerId: AllPeerIds]
+        \union
+    [type: {"syncTimeout"}]
+
+\* All messages (and events) received by a node
+InMsgs ==
+    {NoMsg}
+        \union
+    [type: {"blockResponse"}, peerId: AllPeerIds, block: BlocksWithNil]
+        \union
+    [type: {"statusResponse"}, peerId: AllPeerIds, height: Heights]
+        \union
+    ControlMsgs
+
+\* Messages sent by a node and received by peers (environment in our case)
+OutMsgs ==
+    {NoMsg}
+        \union
+    [type: {"statusRequest"}]           \* StatusRequest is broadcast to the set of connected peers.
+        \union
+    [type: {"blockRequest"}, peerId: AllPeerIds, height: Heights]
+
+
+(********************************** NODE ***********************************)
+
+InitNode ==
+     \E pIds \in SUBSET AllPeerIds \ {{}}:                   \* set of peers node established initial connections with
+        /\ blockPool = [
+                height |-> 1,
+                peerIds |-> pIds,
+                peerHeights |-> [p \in AllPeerIds |-> NilHeight],     \* no peer height is known
+                blockStore |-> [h \in Heights |-> NilBlock],
+                receivedBlocks |-> [h \in Heights |-> NilPeer],
+                pendingBlocks |-> [h \in Heights |-> NilPeer],
+                syncedBlocks |-> -1,
+                syncHeight |-> 1
+           ]
+       /\ state = "running"
+
+\* Remove faulty peers.
+\* Returns new block pool.
+\* See https://github.com/tendermint/tendermint/blob/dac030d6daf4d3e066d84275911128856838af4e/blockchain/v2/scheduler.go#L222
+RemovePeers(peers, bPool) ==
+    LET pIds == bPool.peerIds \ peers IN
+    LET pHeights ==
+        [p \in AllPeerIds |-> IF p \in peers THEN NilHeight ELSE bPool.peerHeights[p]] IN
+
+    LET failedRequests ==
+        {h \in Heights: /\ h >= bPool.height
+                        /\ bPool.pendingBlocks[h] \in peers
+                        /\ bPool.receivedBlocks[h] \in peers} IN
+    LET pBlocks ==
+        [h \in Heights |-> IF h \in failedRequests THEN NilPeer ELSE bPool.pendingBlocks[h]] IN
+    LET rBlocks ==
+        [h \in Heights |-> IF h \in failedRequests THEN NilPeer ELSE bPool.receivedBlocks[h]] IN
+    LET bStore ==
+        [h \in Heights |-> IF h \in failedRequests THEN NilBlock ELSE bPool.blockStore[h]] IN
+
+    IF pIds /= bPool.peerIds
```

Comment:
We can, it would just be a bit more ugly as we would inline definitions in ELSE branch. I thought that TLC doesn't care. @igor Can you please confirm this? Is good practice using LET only inside the path where it is used, or we can extract it to the beginning of the function and then it will just be evaluated if that branch is evaluated.

Reply:
Unless Markus has introduced some optimizations, TLC should evaluate a LET-IN expression, as soon as it meets one. (Although I can imagine it would not be hard to make evaluation lazy.) A common trick in TLC is to cache the often-used constant expressions by defining them in the top-level LET-IN blocks (without arguments).

Reply:
That said, it would be probably more efficient to keep the LET-IN expressions close to the points of their use. However, if it affects readability, I don't think it is a big deal to move them a bit upwards in this case.